File: cloudflare-worker/src/push-provider-senders/discord.ts
import type { NotificationMessage, ProviderConfig, PushResult } from "../types";
import { fetchWithTimeout } from "../rocom-client";
import { providerErrorResult, readResponsePayload, resultFromParsedResponse } from "../push-http";

export async function sendDiscord(
  provider: ProviderConfig,
  message: NotificationMessage,
  timeoutSec: number
): Promise<PushResult> {
  const webhook = provider.config.webhook;
  const separator = webhook.includes("?") ? "&" : "?";

  try {
    const resp = await fetchWithTimeout(
      `${webhook}${separator}wait=true`,
      {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          content: `${message.title}\n\n${message.markdown}`,
          allowed_mentions: { parse: [] },
        }),
      },
      timeoutSec
    );

    const { payload, text } = await readResponsePayload(resp);
    return resultFromParsedResponse(
      provider,
      resp,
      payload,
      text,
      new Set([null, undefined])
    );
  } catch (err) {
    return providerErrorResult(provider, err);
  }
}

File: src/roco_serverchan_notifier/push_provider_senders/discord.py
from __future__ import annotations

import json

import requests

from ..push_http import HttpSession, result_from_status
from ..push_models import NotificationMessage, ProviderConfig, PushResult


def send_discord(
    provider: ProviderConfig, message: NotificationMessage, session: HttpSession, timeout: int
) -> PushResult:
    webhook = str(provider.config["webhook"])
    separator = "&" if "?" in webhook else "?"

    # If image data is available, send as file attachment
    if message.image_data:
        files = {
            "files[0]": ("merchant_card.png", message.image_data, "image/png"),
        }
        data = {
            "payload_json": json.dumps({
                "content": f"{message.title}\n\n{message.body}",
                "allowed_mentions": {"parse": []},
            }),
        }
        try:
            response = session.post(
                f"{webhook}{separator}wait=true",
                data=data,
                files=files,
                timeout=timeout,
            )
            result = result_from_status(provider, response)
            if result.success:
                return result
        except requests.RequestException:
            pass  # Image delivery is optional; still attempt the original text message.

    response = session.post(
        f"{webhook}{separator}wait=true",
        json={
            "content": f"{message.title}\n\n{message.markdown}",
            "allowed_mentions": {"parse": []},
        },
        timeout=timeout,
    )
    return result_from_status(provider, response)
